Target Mode docs (#4267)
chef/chef-web-docs: content/target_mode.md

+++
title = "Target Mode"
draft = false
gh_repo = "chef-web-docs"

[menu]
  [menu.infra]
    title = "Target Mode"
    identifier = "chef_infra/features/Target Mode"
    parent = "chef_infra/features"
    weight = 80
+++

Target Mode executes Chef Infra Client runs on nodes that don't have Chef Infra Client installed on them.

The target node can be any remote system, edge device, or cloud resource that the host can reach. This includes edge devices, Wi-Fi routers, switches, relays, cloud resources, IP phones, router hubs, and network management peripherals.

## Transport Interface (Train)

Target Mode uses [Transport Interface (Train)](https://github.com/inspec/train) to connect to nodes and execute Chef Infra Client runs.

Target Mode supports the SSH Train protocol. The other Train protocols are experimental.

## Requirements

Target Mode has the following requirements:

- A network-enabled system to execute Target Mode.
- The `chef-client` CLI. This is included with Chef Workstation.
- A credentials file which provides the system with information to connect to a target node.
- A recipe that only includes Target Mode-enabled resources.

## Credentials file

> The credentials file lists the nodes that Target Mode can connect to and the configuration for connecting to each node in TOML format.

Author: ```suggestion
The credentials file defines the connection settings for each node in TOML format.
```

The credentials file is located in `.chef/credentials` on Linux and Mac systems, or `c:\Users\<USERNAME>\.chef/credentials` on Windows.

### Examples

Define the list of nodes in the credentials file using the TOML format.
The connection settings for each node are defined using a [TOML Inline Table](https://toml.io/en/v1.0.0#inline-table).

For example, this adds credentials for three nodes using SSH:

```toml
['HOST-1']
host = 'target.system.host.1.com'
user = 'root'
key_files = '~/.keys/key-pair.pem'

['HOST-2']
host = 'target.system.host.2.com'
user = 'root'
password = '123456'

['192.168.0.252']
host = '192.168.0.252'
user = 'root'
password = '123456'

> protocol = ssh

Reviewer: This should be in single quotes. Otherwise it will give syntax error.

> protocol = ssh

Reviewer: Another thing which I want to keep as note here that protocol keyword does not work if using winrm as transport protocol. What works for both ssh and winrm is `transport_protocol`

Reviewer 2: As per Thomas H windows will not be supported for target mode. We will have to make that change later .  You can make the change mentioned above though. @tpowell-progress

### Target Mode in Local Mode

You can run Target Mode in [Local Mode]({{< relref "/ctl_chef_client#run-in-local-mode" >}}).
Local Mode runs chef-zero locally as a lightweight instance of Chef Infra Server to execute a Client run on target nodes.

Use `-z` and `-t` to run Target Mode in Local Mode:

```sh
chef-client -z -t <TARGET_NAME>
```

Replace `<TARGET_NAME>` with the name of the host as defined in the credentials file. For example, `HOST-1` in the [credential file example](#examples).

## Run Target Mode with Chef Automate or Chef Infra Server

You can configure Chef Automate or Chef Infra Server to run Target Mode on a regular schedule.

Target Mode doesn't have a way to schedule Chef Infra Client runs on a node, but you can create a cron file that executes Target Mode on a regular schedule.

For example, this create a cron file that executes Target Mode every thirty minutes:

```ruby
cat > /etc/cron.d/nodename.cron <<EOF
PATH="/opt/chefdk/bin:/opt/chefdk/embedded/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
GEM_ROOT="/opt/chefdk/embedded/lib/ruby/gems/2.6.0"

> 0,30 * * * * * chef-runner /usr/bin/chef-client -t <TARGET_NAME>

Reviewer: As per my knowledge there is  missing peace of information here. 
1. Where this cron file should be? On node or on host?
2. Second when chef-client runs without --local-mode i.e -z option it will check for the client.pem file and client.rb file. That peace of information is missing here. Unless you have client.pem and client.rb file on the host the chef-client run will fail if not running in a local mode.

Reviewer 3: @Vasu1105 Given target mode, I would assume the host is the only one with chef-client, and the usual requirements for chef-client setup would apply. It might be worth calling that out specifically still.
